drivers: i2c: Add I2C driver of it51xxx

GTLin08 opened the pull request:
Implement the functions of I2C host and target.
I2CM: supports nine hosts and each one able located at I2C interface 0-12.
          supports two 32 bytes dedicated FIFO mode for read and write.

I2CS: supports three targets and each one able located at I2C interface 0-8.
         supports 16 bytes dedicated FIFO mode that only supports write or read mode and the maximum buffer size is 256 bytes.
         support non-FIFO write to shared FIFO read mode. The maximum shared FIFO size for read is 256 bytes.

The APIs test include: i2c_write(), i2c_read(), i2c_burst_read(), i2c_burst_write(), i2c_write_read()

GTLin08 committed:
drivers/i2c: Add I2C driver of it51xxx

Implement the functions of I2C host and target.
I2CM: supports nine hosts and each one able located at I2C interface
      0~12.
      supports two 32 bytes dedicated FIFO mode for read and write.
I2CS: supports three targets and each one able located at I2C
      interface 0~8.
      supports 16 bytes dedicated FIFO mode that only supports write or
      read mode and the maximum buffer size is 256 bytes.
      support non-FIFO write to shared FIFO read mode. The maximum
      shared FIFO size for read is 256 bytes.
The APIs test include: i2c_write(), i2c_read(), i2c_burst_read(),
                       i2c_burst_write(), i2c_write_read()

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 committed:
test: drivers/i2c: ite: it515xx_evb: Add I2C target test

Add I2C target node for it515xx_evb test.

Signed-off-by: Tim Lin <tim2.lin@ite.corp-partner.google.com>

GTLin08 commented:
Rebased.

kartben reviewed (dismissed):
Coding guidelines issue pointed out by Copilot should be addressed. Thanks!

GTLin08 commented on soc/ite/ec/it51xxx/chip_chipregs.h:
```diff
@@ -321,4 +321,16 @@ struct gctrl_it51xxx_regs {
+/* 0x0a, 0x2a, 0x4a: Slave n Dedicated FIFO Pre-defined Control Register */
+#define SMB_SADFPCTL     IT51XXX_SMB_BASE + 0x0a
+#define SMB_SBDFPCTL     IT51XXX_SMB_BASE + 0x2a
+#define SMB_SCDFPCTL     IT51XXX_SMB_BASE + 0x4a
```
Done.

GTLin08 commented on dts/bindings/i2c/ite,it51xxx-i2c.yaml:
```diff
@@ -0,0 +1,201 @@
+      10 = SMB_SWITCH_INTERFACE9: Swich to interface9 SMCLK9/SMDAT9
+      11 = SMB_SWITCH_INTERFACE10: Swich to interface10 SMCLK10/SMDAT10
+      12 = SMB_SWITCH_INTERFACE11: Swich to interface11 SMCLK11/SMDAT11
+      13 = SMB_SWITCH_INTERFACE12: Swich to interface12 SMCLK12/SMDAT12
```
Done.

GTLin08 commented:
Hi @teburd, @Dino-Li, @kartben,
I’ve made changes based on Copilot’s requests. Please revisit the updated PR. Thanks!

GTLin08 commented:
Hi @Dino-Li ,

Rebased and fixed the issue as described below:

While working with I2C3, we discovered an issue in the previous PR where the reg property was incorrectly defined as
```
i2c3: i2c@f04178 {
     reg = <0x00f04178 0x0028>;
}
```
This should be:
```
i2c3: i2c@f04178 {
     reg = <0x00f04178 0x0028
            0 1>;
}
```
I've corrected this in the latest update.

GTLin08 commented:
Hi @teburd ,

Any comments would be greatly appreciated.

teburd reviewed (approved):
lgtm, reapproving
